### PR TITLE
More secure path mangling

### PR DESCRIPTION
**Pull request by marcjansen**
This is a follow up from the excellent #280. I think the security might even be improved by the following PR.

I'd really like a review from @anikaweinmann or @mmacata on this; this is open for discussion.

### CODE REVIEW

**Comment by marcjansen** [comment]
Hmm. Probably some occasions missed. Hopefully just that.

File: src/actinia_core/core/utils.py [hunk 1]
```diff
@@ -49,11 +50,22 @@ def os_path_normpath(path_parts):
...
         raise SecurityError(f"Used path '{path}' is insecure")
     return path
 
 
+def path_is_in_allowed_paths(path):
+    # TODO we should double check whether they are all needed, better would be one
+    # single location
+    allowed_starts = [
+        global_config.GRASS_USER_DATABASE,
+        global_config.GRASS_DATABASE,
```
Comment (anikaweinmann) [inline src/actinia_core/core/utils.py:63]: Carmen is right that `GRASS_DATABASE` is only okay if read and not if you want to change something in it. Maybe you should still distinguish that somehow?

File: src/actinia_core/core/utils.py [hunk 2]
```diff
@@ -49,11 +50,22 @@ def os_path_normpath(path_parts):
...
+        global_config.GRASS_RESOURCE_DIR
+    ]
+    return any(path.startswith(allowed) for allowed in allowed_starts)
```
Comment (anikaweinmann) [inline src/actinia_core/core/utils.py:66]: Theoretically the `startswith` could be unsafe here again, if  `path="GRASS_RESOURCE_DIR/../../something_else`.

File: src/actinia_core/core/utils.py (hunk 1 above)
Comment (marcjansen) [reply src/actinia_core/core/utils.py:63]: Well it is used in [these lines of PR 280](https://github.com/mundialis/actinia_core/pull/280/files#diff-ee1d2bbb5ac0ad6fc98c8c9b1f98cb33232ee8d70320349b17362de44c629edcR217), that is why I included it.

File: src/actinia_core/core/utils.py (hunk 2 above)
Comment (anikaweinmann) [reply src/actinia_core/core/utils.py:66]: `os.path.normpath("GRASS_RESOURCE_DIR/../../something_else")` leads to `../something_else`, so the normpath is needed that this function `path_is_in_allowed_paths` is working. Maybe you can mention that path should be a path which was corrected with `os.path.normpath`.

File: src/actinia_core/core/utils.py (hunk 1 above)
Comment (anikaweinmann) [reply src/actinia_core/core/utils.py:63]: I am not sure if it makes sense to add some `r`/`w` parameter to the functions and split the `allowed_starts`.

Comment (marcjansen) [reply src/actinia_core/core/utils.py:63]: Ah, I now understand what you mean. I can update this as suggested.

File: src/actinia_core/core/utils.py (hunk 2 above)
Comment (marcjansen) [reply src/actinia_core/core/utils.py:66]: But if `os.path.normpath("GRASS_RESOURCE_DIR/../../something_else")` leads to `../something_else` the function would return false, as it does not start with any of the allowed prefixes. I can of course add a note

**Comment by anikaweinmann** [comment]
@marcjansen do you want to add a `r`/`w` parameter to the functions? Otherwise it is also fine for me to merge it as it is.

**Review by anikaweinmann — APPROVED** [review]

**Comment by marcjansen** [comment]
Sorry for not giving this more attention. I think the general idea is still valid, but I am not sure whether this catches all relevant cases. 

I see this has an approving review, and I can try to rebase this, if you still want it. I can also live with you guys and girls closing this PR with unmerged commits. Please go ahead as you see fit.

**Comment by anikaweinmann** [comment]
@marcjansen you are welcome to merge and rebase

File: src/actinia_core/core/utils.py (hunk 1 above)
Comment (marcjansen) [reply src/actinia_core/core/utils.py:109]: I have updated the PR; `ensure_valid_path` now accepts an optional second parameter, the intent ("r", "w" or "rw", Default is "r")

**Comment by marcjansen** [comment]
> @marcjansen you are welcome to merge and rebase

I have done that and also added support for intents (`r`, `w` or `rw`) when `ensure_valid_path` is called.

Please have another close look at what is now in this PR. Where the method is called, I have added an intent where I was more or less sure, but these shpould please be double checked by someone with more know-how of the actual intent.

**Comment by marcjansen** [comment]
anybody @mundialis-dev ?

**Comment by marcjansen** [comment]
Friendly ping @mmacata @anikaweinmann

**Comment by anikaweinmann** [comment]
Thank you, for me it look good. Please merge.

**Comment by marcjansen** [comment]
Thanks for the review.

**Pull request merged**